shell/common/asar/archive_win.cc

```diff
@@ -0,0 +1,179 @@
+                                  INTEGRITY_CHECK_RESOURCE_TYPE);
+  if (!resource) {
+    LOG(FATAL) << "FindResource failed. Last error: " << GetLastError();
+    return;
+  }
+
+  HGLOBAL rcData = ::LoadResource(module_handle, resource);
+  if (!rcData) {
+    LOG(FATAL) << "LoadResource failed. Last error: " << GetLastError();
+    return;
+  }
+
+  auto* res_data = static_cast<const char*>(::LockResource(rcData));
+  int res_size = SizeofResource(NULL, resource);
+
+  if (!res_data || !res_size) {
+    LOG(FATAL) << "Failed to integrity config from exe resource.";
+    return;
+  }
+
+  // Parse integrity config payload
+  auto integrity_config_payload = string(res_data, res_size);
+  absl::optional<base::Value> root = base::JSONReader::Read(
+      integrity_config_payload,
+      base::JSON_PARSE_CHROMIUM_EXTENSIONS | base::JSON_ALLOW_TRAILING_COMMAS);
+
+  if (!root.has_value()) {
+    LOG(FATAL) << "Invalid integrity config: NOT a valid JSON.";
+    return;
+  }
+
+  const base::Value::List* file_configs = root.value().GetIfList();
+  if (!file_configs) {
+    LOG(FATAL) << "Invalid integrity config: NOT a list.";
+    return;
+  }
+
+  // Parse each individual file integrity config
+  for (size_t i = 0; i < file_configs->size(); i++) {
+    // Skip invalid file configs
+    const Value::Dict* ele_dict = (*file_configs)[i].GetIfDict();
+    if (!ele_dict) {
+      LOG(WARNING) << "Skip config " << i << ": NOT a valid dict";
+      continue;
+    }
+
+    const string* file = ele_dict->FindString("file");
+    if (!file || file->empty()) {
+      LOG(WARNING) << "Skip config " << i << ": Invalid file";
+      continue;
+    }
+
+    const string* alg = ele_dict->FindString("alg");
+    if (!alg || base::ToLowerASCII(*alg) != "sha256") {
+      LOG(WARNING) << "Skip config " << i << ": Invalid alg";
+      continue;
+    }
+
+    const string* value = ele_dict->FindString("value");
+    if (!value || value->empty()) {
+      LOG(WARNING) << "Skip config " << i << ": Invalid hash value";
+      continue;
+    }
+
+    // Add valid file config into cache
+    IntegrityPayload header_integrity;
+    header_integrity.algorithm = HashAlgorithm::kSHA256;
+    header_integrity.hash = *value;
+
+    s_win_integrity_payloads_cache->value()[ToLowerASCII(*file)] =
+        header_integrity;
+  }
+}
+
+absl::optional<IntegrityPayload> Archive::HeaderIntegrity() const {
+  absl::optional<base::FilePath> relative_path = RelativePath();
+  // Callers should have already asserted this
+  CHECK(relative_path.has_value());
+
+  // Load integrity config from exe resource
+  LoadIntegrityConfigCache();
+  if (!s_win_integrity_payloads_cache->has_value()) {
+    LOG(WARNING) << "Failed to integrity config from exe resource.";
+    return absl::nullopt;
+  }
+
+  // Convert Window rel path to ASCII lower case
+  string rel_path_utf8 = WideToASCII(relative_path.value().value().c_str());
```

Comment:
Are we guaranteed that the relative path is ASCII? `base::WideToUTF8` might be more appropriate here. If it really is important that it be ASCII, this fact deserves a comment.

Also, `.c_str()` isn't required as `base::WideTo*` works on `std::wstring_view`.
```suggestion
  string rel_path_utf8 = base::WideToUTF8(relative_path.value().value());
```

Reply:
You are correct. I should convert to UTF8. Updated.